### ISSUE TITLE
Support for objects?

### ISSUE THREAD

**Issue by jlem** [ramsey/array_column#2]
It would be great if you could pass in an array of objects, and the index key provided merely accessed one of the properties to be used when re-indexing the array of objects.

I know that you can implement the ArrayAccess interface, but it's a bit tedious to do this for every object that you just want to be able to reindex an array of. 

A simple check to see if `$row` is an object would be good. A more robust implementation might allow you to define a method call against a class as the index key e.g. `array('Person', 'getID')`.

Of course this would make it inconsistent with the actual PHP implementation, but it would make this more useful. Having to implement ArrayAccess marginally defeats the purpose of using the terse syntax of `array_column()` over the slightly more verbose `array_map()`

**Comment by vlakoff** [comment]
I just noticed PHP 7 [will support array of objects](https://github.com/php/php-src/blob/PHP-7.0/UPGRADING) :-)

**Comment by ramsey** [comment]
Interesting. So, if I back-port it into this library, the object support will only work for `array_column()` in PHP versions 5.4 and earlier, since this library does not override the behavior of the function in 5.5 and 5.6. So, if someone uses this library on 5.4 with objects, when they upgrade to 5.5 or 5.6, their code will stop working.

Thoughts?

**Comment by CrandellWS** [comment]
Create a feature override branch that does override and implements newer behaviors there and use that to create a secondary script perhaps array_column_7. 

I know a single script is probably prefered but you just can not override and default to core at the same time. I feel it best used as a plug and play type setup(no configs). You have the script it works just choose your preference array_column or array_column_7 ...

**Comment by ramsey** [comment]
I'd probably namespace it, rather than use `array_column_7`. Maybe `Ramsey\Polyfill\array_column()`. I'll give it some thought. Thanks!

**Comment by vlakoff** [comment]
Personally I had added an `object_column()` helper in my codebases. This way, arrays and objects are clearly separated, which has pros and cons from an API perspective. At least it has some performance benefit, as the item types don't have to be checked.

Your proposals sound good too, but I can think of some caveats. For example, you could forget to use the namespaced function and it would go unnoticed, the code would work just fine until an object is encountered or you switch from PHP 7 to 5.

**Comment by vlakoff** [comment]
I would suggest creating a new package "object_column", defining a function `object_column()` accepting only objects and working with them like `array_column()` of PHP 7.

Once the user has moved to PHP 7 and doesn't plan to go back, he could simply replace `object_column()` with `array_column()` in the codebase to use the native function.

**Comment by vlakoff** [comment]
And on second thought, -1 on the namespaced function proposal. I think it would be far too prone to forget the import and use the native function by mistake.

**Comment by ramsey** [comment]
Yeah. I'm not really sure the best approach here. This polyfill is for people using PHP 5.4 and earlier. If anyone is using 5.5 or 5.6, object support in this package won't help them, since they'll be using the native PHP function. I assume people are more likely to go from 5.3 and 5.4 to 5.6 before moving to 7.0, which will cause their applications to break, if they rely on this library with object support.

**Comment by vlakoff** [comment]
That's precisely why I thought about creating a new package, bringing a separate function `object_column` which would complement `array_column` without having to care if it is the native PHP function or the polyfill :)

**Comment by shadowhand** [comment]
PHP 7.0 supports objects. Please update this package for those of us stuck on PHP < 7 or HHVM!

**Comment by shadowhand** [comment]
Though as it turns out, this won't help us, because PHP/HHVM already defined `array_column` and so this package will not provide a 7.x compatible function. 😭

**Comment by ramsey** [comment]
Trying to provide compatibility to previous versions when the function changes drastically across versions in which it is implement is hard.

**Comment by joshfoskett** [comment]
It says that the ability to pass an array of objects as `input` was added in `7.0`, but it also works for me in `5.6.10`. I'm confused.

**Comment by iisisrael** [comment]
Another object-related backwards compatibility issue to keep in mind is that if an object is the value of the column, it will be string cast.  I don't know if that is a PHP 7.0 bug or not, but I just ran into that today.  Anyone implementing this package and depending on an array of objects being returned will have any number of runtime errors when they upgrade PHP.

```
class MyObject {
    public function __toString()
    {
        return 'I am a string';
    }
}

$myArray = array(
    array('one' => new MyObject(), 'two' => new MyObject()),
    array('one' => new MyObject(), 'two' => new MyObject())
);

$twoArray = array_column($myArray, 'two');
```
Using this package, `$twoArray` will be an array of two objects, whereas with PHP 7.0 `$twoArray` will be an array of two "I am a string" strings.
